### PR TITLE
Rework test suite to depend less heavily on mocks and implementation.

### PR DESCRIPTION
**Pull request by collijk**
### Description

- *Category*: refactor
- *JIRA issue*: [MIC-3068](https://jira.ihme.washington.edu/browse/MIC-3068)

The existing test suite made very heavy use of mock objects and was 
primarily checking implementation details.  This is generally not a great
method for writing test code as it is both very brittle (implementation 
details shouldn't be contractually enforced, only the interface) and does
not actually check the behavior.

This PR reworks the existing test suite to build interactive simulations
using existing testing utilities and tests public behavior and contracts
in actual use case settings.

### Testing
Only test code in this PR

### CODE REVIEW

File: tests/population/test_add_new_birth_cohort.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def crude_birth_rate_data(live_births=500):
             ("age", "year", "sex", "parameter", "value"),
         )
         .query('age_start == 25 and sex != "Both"')
-        .drop(["age_start", "age_end"], "columns")
```
Comment (collijk) [inline tests/population/test_add_new_birth_cohort.py:46]: Forgot one in the previous pr

File: tests/population/test_base_population.py [hunk 2]
```diff
@@ -291,14 +291,14 @@ def test__assign_demography_with_initial_age_error(include_sex):
         )
 
 
-def test__assign_demography_with_age_bounds(include_sex):
```
Comment (collijk) [inline tests/population/test_base_population.py:294]: All the population tests are just adjustments to new non-uniform test age bins.

**Review by collijk — COMMENTED** [review]
All the important changes are in test_stratification.py

File: tests/metrics/test_stratification.py [hunk 3]
```diff
@@ -1,371 +1,252 @@
...
 from typing import Callable, List, Set
 from unittest.mock import MagicMock, patch
 
+import numpy as np
 import pandas as pd
 import pytest
+from vivarium import InteractiveContext
+from vivarium.framework.engine import Builder
+from vivarium.framework.population import SimulantData
+from vivarium.testing_utilities import TestPopulation, metadata
```
Comment (collijk) [inline tests/metrics/test_stratification.py:13]: Vivarium and vivarium public health both have a set of test utilities.  Vivarium provides a base population component and some tools for manipulation configuration metadata.  Vivarium public health has a mock artifact manager that supplies test data (see `vivarium_public_health.testing.mock_artifact` for details).

File: tests/metrics/test_stratification.py [hunk 4]
```diff
@@ -1,371 +1,252 @@
...
-    ResultsStratifier,
+    ResultsStratifier as ResultsStratifier_,
+)
+from vivarium_public_health.metrics.stratification import (
     Source,
     SourceType,
     StratificationLevel,
 )
 
-AGE_BINS = pd.DataFrame(
```
Comment (collijk) [inline tests/metrics/test_stratification.py:16]: The purpose of all those test utilities is to make it so you don't need to write a ton of mock code to test new components like what's here.

File: tests/metrics/test_stratification.py [hunk 5]
```diff
@@ -1,371 +1,252 @@
...
-    "month": StratificationLevel(
-        name="month",
-        sources=[Source("clock_2", SourceType.CLOCK)],
-        categories={str(i + 1) for i in range(12)},
-        mapper=lambda row: str(row[0].month),
-        current_categories_getter=lambda: {str(mock_clock().month)},
-    ),
-}
 
+class FavoriteColor:
```
Comment (collijk) [inline tests/metrics/test_stratification.py:25]: I'm also shifting strategies.  Rajan's test were all "white box" tests which had intimate knowledge of the stratifier component internals.  What I'm doing instead is "black box" testing, which doesn't know anything about the component beyond it's public interface and contracts. I'm doing this by generating minimal interactive simulations with all the features necessary to stress test the results stratifier and then testing the component interface.  This embeds the new component in a stripped down, but realistic setting where it's much easier to check for expected behavior.

File: tests/metrics/test_stratification.py [hunk 6]
```diff
@@ -1,371 +1,252 @@
...
+FAVORITE_THINGS = {
+    f"{color}_{number}"
+    for color, number in itertools.product(FavoriteColor.OPTIONS, FavoriteNumber.OPTIONS)
+}
 
-def mock_clock() -> datetime:
-    return datetime(2022, 5, 3, 12)
 
+class ResultsStratifier(ResultsStratifier_):
+    def register_stratifications(self, builder: Builder) -> None:
```
Comment (collijk) [inline tests/metrics/test_stratification.py:78]: Here is another example. We are doing what we'd do in actual practice -- extending the results stratifier to add new startifications.  Test suites can often serve as example use cases when written well.

File: src/vivarium_public_health/testing/utils.py [hunk 7]
```diff
@@ -30,9 +31,41 @@ def make_uniform_pop_data(age_bin_midpoint=False):
...
     )
     if age_bin_midpoint:  # used for population tests
         pop["age"] = pop.apply(lambda row: (row["age_start"] + row["age_end"]) / 2, axis=1)
     return pop
+
+
+def make_age_bins():
+    idx = pd.MultiIndex.from_tuples(
+        [
+            (0.0, 0.01917808, "Early Neonatal"),
```
Comment (albrja) [inline src/vivarium_public_health/testing/utils.py:45]: We actually ran into this yesterday. Is there a reason we do not include the "birth" age group in this?

**Review by albrja — APPROVED** [review]

**Review by mattkappel — APPROVED** [review]

**Pull request merged**